Fix old registration API detection

pevik opened the pull request:
Reliably detect whether functions from old API exists.
    
Depending on SLES specific KLP_NOREG_API does not work on Tumbleweed and mainline.
    
Fixes: f5e7bad Adapt the livepatch template to a new livepatch API

pevik committed:
Simplify detection of kernel build directory

/lib/modules/$(uname -r)/build is a symlink to
/usr/src/linux-$KERN_VERSION-obj/$KERN_ARCH/$KERN_FLAVOR on common
linux distros.

Signed-off-by: Petr Vorel <pvorel@suse.cz>

pevik committed:
Fix old registration API detection

Depending on SLES specific KLP_NOREG_API does not work on Tumbleweed
and mainline. So also check kernel version (simplified KLP API was
merged as 958ef1e39d24 in 5.1.0).

NOTE: this still can fail for other distros which backport simplified
API to older kernel versions. IMHO grep klp_register_patch
/lib/modules/$(uname -r)/source/include/linux/livepatch.h is the most
reliable way how to detect which API is used.  (+ we could also employ
autotools to detect, but that's too big overhead).

Fixes: f5e7bad Adapt the livepatch template to a new livepatch API

Signed-off-by: Petr Vorel <pvorel@suse.cz>

pevik committed:
Remove duplicity of creating header

Signed-off-by: Petr Vorel <pvorel@suse.cz>

pevik commented:
@mimi1vx @mdoucha @czerw: FYI

pevik commented:
NOTE: if you don't like path change (first commit), I can provide version without it.

lpechacek commented:
Thanks, Petr, for the patch set! The path change is fine with me and I like it.
However, I do have a few comments on the last patch. I'll post them inline.

lpechacek commented on klp_tc_functions.sh:
```diff
@@ -28,7 +28,8 @@ function __klp_add_patched_func() {
 function klp_create_header() {
-    sed "s%@@USE_OLD_HRTIMER_API@@%$KLP_TEST_HRTIMER_OLD%" \
+    sed -e "s%@@USE_OLD_HRTIMER_API@@%$KLP_TEST_HRTIMER_OLD%" \
+	    -e "s%@@USE_OLD_REG_API@@%$KLP_TEST_REG_API%" \
```
`USE_OLD_REG_API` macro is used only in `klp_test_livepatch.c`. Therefore I'd rather see the change made to that file and not the support mod's header file.

lpechacek commented on klp_tc_functions.sh:
```diff
@@ -311,6 +312,14 @@ if [ ! -f $KLP_ENV_CACHE_FILE ]; then
     fi
 
+    echo -n 'export KLP_TEST_REG_API=' >> $KLP_ENV_CACHE_FILE
+    if grep klp_register_patch "/lib/modules/$(uname -r)/source/include/linux/livepatch.h"
```
The kernel header file is yet another loose part in the system. Of course, without it the test modules won't compile, but I'd still prefer making the decision based on information provided by running kernel image. I suggest either grepping `/proc/kallsyms` or checking kernel version. Cf. https://github.com/lpechacek/qa_test_klp/blob/19692b48765a936996f1e7775700c62de1c4bfeb/klp_tc_functions.sh#L336-L339

pevik replied:
Sure :). I found this commit [958ef1e39d24 livepatch: Simplify API by removing registration step](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=958ef1e39d24d6cb8bf2a7406130a98c9564230f) merged into v5.1-rc1, so I'll `$VERSION_CODE` for it.

pevik replied:
Uh, took me a while to realize that you offered code to copy paste :)

pevik replied:
@lpechacek Libor, this will not work for SLE12 SP4 (old version, but backported patches). I'll also add check about `KLP_NOREG_API` not being defined. But IMHO grepping headers is more reliable for all distros.

lpechacek commented on klp_test_support_mod.h:
```diff
@@ -21,6 +21,7 @@
 #include <asm/uaccess.h>
 
 #define USE_OLD_HRTIMER_API @@USE_OLD_HRTIMER_API@@
+#define USE_OLD_REG_API @@USE_OLD_REG_API@@
```
Should go to `klp_test_livepatch.c`.

pevik committed:
Generate linux version portable way

mapfile needed -d, otherwise:
./klp_tc_functions.sh: line 320: 5 4 0: syntax error in expression (error token is "4 0")

and mapfile is a bashism anyway :)

Fixes: 2827922 Fix 'mapfile: -d: invalid option' error

Signed-off-by: Petr Vorel <pvorel@suse.cz>

pevik commented:
@lpechacek Also added fix for 2827922, see http://quasar.suse.cz/tests/4226/file/serial_terminal.txt
`./klp_tc_functions.sh: line 320: 5 3 13: syntax error in expression (error token is "3 13")`

IMHO POSIX shell syntax is safer than bashisms to avoid these problems.

pevik commented:
Ready. 
http://quasar.suse.cz/tests/4230/file/serial_terminal.txt
I just don't like negative `#if !USE_NOREG_API` (could be `#if USE_OLD_REG_API` with opposite value), but that's a detail.

mdoucha commented:
QAM verification runs: https://openqa.suse.de/tests/overview?distri=sle&build=klp_update_test

pevik commented:
Added fix for implicit declaration errors on SLE12 SP5 (old kernel with backport)
https://openqa.suse.de/tests/3675970/file/serial_terminal.txt

Fixed as `#if !(defined(KLP_NOREG_API) || USE_NOREG_API)`, but IMHO greping headers is better, because it works for all distros (including these which backport new API).

pevik commented:
All running versions are working with current version: https://openqa.suse.de/tests/overview?distri=sle&build=klp_update_test
With that grep header vs. `#if !(defined(KLP_NOREG_API) || USE_NOREG_API)` it's up to you, what you prefer.

lpechacek commented:
Patches from this pull request have been reworked in `use-klp-simplified-api` branch and merged into `master`. Thanks for the proposal and smooth cooperation!
